src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -0,0 +1,722 @@
+        case (TSet(et1, r1), TSet(et2, r2)) => r1 == r2 && isSupertype(et1, et2)
+        case (TDict(kt1, vt1, r1), TDict(kt2, vt2, r2)) => r1 == r2 && isSupertype(kt1, kt2) && isSupertype(vt1, vt2)
+        case (s1: TStruct, s2: TStruct) =>
+          var idx = -1
+          s1.required == s2.required && s1.fields.forall { f =>
+            val s2field = s2.field(f.name)
+            if (s2field.index > idx) {
+              idx = s2field.index
+              isSupertype(f.typ, s2field.typ)
+            } else
+              false
+          }
+        case (t1: TTuple, t2: TTuple) =>
+          t1.required == t2.required &&
+            t1.size == t2.size &&
+            t1.types.zip(t2.types)
+              .forall { case (elt1, elt2) => isSupertype(elt1, elt2) }
+        case (agg1: TAggregable, agg2: TAggregable) => isSupertype(agg1.elementType, agg2.elementType)
+        case (t1: Type, t2: Type) => t1 == t2
+        case _ => fatal(s"invalid comparison: $superType / $subType")
+      }
+    } catch {
+      case e: Throwable =>
+        fatal(s"error while checking subtype:\n  super: $superType\n  sub:   $subType", e)
+    }
+  }
+
+  def apply(ir: BaseIR): BaseIR = {
+
+    try {
+      val irCopy = ir.deepCopy()
+      val memo = Memo.empty[BaseType]
+      irCopy match {
+        case mir: MatrixIR =>
+          memoizeMatrixIR(mir, mir.typ, memo)
+          rebuild(mir, memo)
+        case tir: TableIR =>
+          memoizeTableIR(tir, tir.typ, memo)
+          rebuild(tir, memo)
+        case vir: IR =>
+          memoizeValueIR(vir, vir.typ, memo)
+          rebuild(vir, Env.empty[Type], memo)
+      }
+    } catch {
+      case e: Throwable => fatal(s"error trying to rebuild IR:\n${ Pretty(ir) }", e)
+    }
+  }
+
+
+  def minimal(tt: TableType): TableType = {
+    val keySet = tt.key.iterator.flatten.toSet
+    tt.copy(
+      rowType = tt.rowType.filter(keySet)._1,
+      globalType = TStruct(tt.globalType.required)
+    )
+  }
+
+  def minimal(mt: MatrixType): MatrixType = {
+    val rowKeySet = mt.rowKey.toSet
+    val colKeySet = mt.colKey.toSet
+    mt.copy(
+      rvRowType = TStruct(mt.rvRowType.required, mt.rvRowType.fields.flatMap { f =>
+        if (rowKeySet.contains(f.name))
+          Some(f.name -> f.typ)
+        else if (f.name == MatrixType.entriesIdentifier)
+          Some(f.name -> minimal(f.typ))
+        else
+          None
+      }: _*),
+      colType = mt.colType.filter(colKeySet)._1,
+      globalType = TStruct(mt.globalType.required)
+    )
+  }
+
+  def minimal[T <: Type](base: T): T = {
+    val result = base match {
+      case ts: TStruct => TStruct(ts.required)
+      case ta: TArray => TArray(minimal(ta.elementType), ta.required)
+      case tagg: TAggregable => tagg.copy(elementType = minimal(tagg.elementType))
+      case t => t
+    }
+    result.asInstanceOf[T]
+  }
+
+  def unifyBaseType(base: BaseType, children: BaseType*): BaseType = {
```

Comment:
Children seems like the wrong name here?  These are subtypes of base that are being unioned together, right?

Reply:
Can we overload this and have a Seq version and a varargs BaseType* version?